File: backend/app/routers/sessions.py
"""会话 CRUD + 回放 API（设计文档 §5）。

路由注册顺序注意：``/search`` 必须在 ``/{session_id}`` 之前，
否则 "search" 会被当作 session_id 匹配。
"""

from __future__ import annotations

import asyncio
import json

from fastapi import APIRouter, HTTPException, Request

from backend.app.deps.user import get_user_id
from backend.app.schemas.session import RenameRequest
from backend.app.services import session_store

router = APIRouter(prefix="/api/sessions", tags=["sessions"])


def _load_owned_session(session_id: str, user_id: str) -> dict:
    """归属校验：不存在 → 404，非本人 → 403（设计文档 §2）。"""
    sess = session_store.get_session(session_id)
    if sess is None:
        raise HTTPException(status_code=404, detail="会话不存在")
    if sess["user_id"] != user_id:
        raise HTTPException(status_code=403, detail="无权访问该会话")
    return sess


@router.post("")
async def create_session(request: Request) -> dict:
    user_id = get_user_id(request)
    sid = await asyncio.to_thread(session_store.create_session, user_id)
    return {"session_id": sid}


@router.get("")
async def list_sessions(request: Request) -> dict:
    user_id = get_user_id(request)
    sessions = await asyncio.to_thread(session_store.list_sessions, user_id)
    return {"sessions": sessions}


@router.get("/search")
async def search_sessions(request: Request, q: str = "") -> dict:
    user_id = get_user_id(request)
    results = await asyncio.to_thread(session_store.search_sessions, user_id, q)
    return {"sessions": results}


@router.get("/{session_id}")
async def replay_session(session_id: str, request: Request) -> dict:
    """回放协议（设计文档 §5）：trace 原样返回，steps/tools 由 trace 推导。"""
    user_id = get_user_id(request)
    sess = await asyncio.to_thread(_load_owned_session, session_id, user_id)
    rows = await asyncio.to_thread(session_store.list_messages, session_id)
    messages = []
    for m in rows:
        trace = json.loads(m["artifacts_json"]) if m["artifacts_json"] else None
        messages.append(
            {
                "id": m["id"],
                "role": m["role"],
                "content": m["content"],
                "steps": len(trace) if trace else None,
                "tools": list(dict.fromkeys(t["tool"] for t in trace)) if trace else None,
                "trace": trace,
                "created_at": m["created_at"],
            }
        )
    return {
        "session": {
            "id": sess["id"],
            "title": sess["title"],
            "created_at": sess["created_at"],
            "updated_at": sess["updated_at"],
        },
        "messages": messages,
    }


@router.patch("/{session_id}")
async def rename_session(session_id: str, body: RenameRequest, request: Request) -> dict:
    user_id = get_user_id(request)
    await asyncio.to_thread(_load_owned_session, session_id, user_id)
    await asyncio.to_thread(session_store.rename_session, session_id, body.title)
    return {"ok": True}


@router.delete("/{session_id}")
async def delete_session(session_id: str, request: Request) -> dict:
    user_id = get_user_id(request)
    await asyncio.to_thread(_load_owned_session, session_id, user_id)
    await asyncio.to_thread(session_store.delete_session, session_id)
    return {"ok": True}

File: frontend/src/stores/sessions.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/utils/api'
import type { ToolTrace } from './chat'

export interface SessionSummary {
  id: string
  title: string | null
  updated_at: string
}

export interface ReplayMessage {
  id: number
  role: 'user' | 'assistant'
  content: string
  steps: number | null
  tools: string[] | null
  trace: ToolTrace[] | null
  created_at: string
}

export interface ReplayResponse {
  session: { id: string; title: string | null; created_at: string; updated_at: string }
  messages: ReplayMessage[]
}

export const useSessionsStore = defineStore('sessions', () => {
  const list = ref<SessionSummary[]>([])
  const currentId = ref<string | null>(null)

  async function fetchList() {
    const data = await api<{ sessions: SessionSummary[] }>('/api/sessions')
    list.value = data.sessions
  }

  async function create(): Promise<string> {
    const data = await api<{ session_id: string }>('/api/sessions', {
      method: 'POST',
      body: '{}',
    })
    currentId.value = data.session_id
    await fetchList()
    return data.session_id
  }

  /** 幂等主流程（设计文档 §5）：发送前调用，无当前会话则先创建。 */
  async function ensureSession(): Promise<string> {
    if (!currentId.value) {
      await create()
    }
    return currentId.value!
  }

  async function rename(id: string, title: string) {
    await api(`/api/sessions/${id}`, { method: 'PATCH', body: JSON.stringify({ title }) })
    await fetchList()
  }

  async function remove(id: string) {
    await api(`/api/sessions/${id}`, { method: 'DELETE' })
    if (currentId.value === id) currentId.value = null
    await fetchList()
  }

  async function search(q: string): Promise<SessionSummary[]> {
    const data = await api<{ sessions: SessionSummary[] }>(
      `/api/sessions/search?q=${encodeURIComponent(q)}`,
    )
    return data.sessions
  }

  async function loadReplay(id: string): Promise<ReplayResponse> {
    return api<ReplayResponse>(`/api/sessions/${id}`)
  }

  return {
    list,
    currentId,
    fetchList,
    create,
    ensureSession,
    rename,
    remove,
    search,
    loadReplay,
  }
})
